src/main/host/descriptor/descriptor_table.rs

```diff
@@ -0,0 +1,204 @@
+            super::export::compatdescriptor_setHandle(&mut prev as *mut CompatDescriptor, 0);
+            Some(prev)
+        } else {
+            trace!("Setting to unused index {}", index);
+            None
+        }
+    }
+
+    /// This is a helper function that handles some corner cases where some
+    /// descriptors are linked to each other and we must remove that link in
+    /// order to ensure that the reference count reaches zero and they are properly
+    /// freed. Otherwise the circular reference will prevent the free operation.
+    /// TODO: remove this once the TCP layer is better designed.
+    pub fn shutdown_helper(&mut self) {
+        for (_idx, descriptor) in self.descriptors.iter() {
+            match descriptor {
+                CompatDescriptor::New(_) => continue,
+                CompatDescriptor::Legacy(d) => unsafe {
+                    cshadow::descriptor_shutdownHelper(d.ptr())
+                },
+            };
+        }
+    }
+}
+
+mod export {
+    use super::*;
+    use libc::c_int;
+
+    /// Create an object that can be used to store all descriptors created by a
+    /// process. When the table is no longer required, use descriptortable_free
+    /// to release the reference.
+    #[no_mangle]
+    pub unsafe extern "C" fn descriptortable_new() -> *mut DescriptorTable {
+        Box::into_raw(Box::new(DescriptorTable::new()))
+    }
+
+    /// Free the table.
+    #[no_mangle]
+    pub unsafe extern "C" fn descriptortable_free(table: *mut DescriptorTable) {
+        unsafe { Box::from_raw(notnull_mut_debug(table)) };
+    }
+
+    /// Store a descriptor object for later reference at the next available index
+    /// in the table. The chosen table index is stored in the descriptor object and
+    /// returned. The descriptor is guaranteed to be stored successfully.
+    ///
+    /// NOTE: that this consumes a reference to the descriptor, so if you are storing
+    /// it outside of the descriptor table you will need to ref it after calling
+    /// this function.
+    #[no_mangle]
+    pub unsafe extern "C" fn descriptortable_add(
+        table: *mut DescriptorTable,
+        descriptor: *mut CompatDescriptor,
+    ) -> c_int {
+        let table = unsafe { table.as_mut().unwrap() };
+        let descriptor = CompatDescriptor::from_raw(descriptor).unwrap();
+        table.add(*descriptor).try_into().unwrap()
+    }
+
+    /// Stop storing the descriptor so that it can no longer be referenced. The table
+    /// index that was used to store the descriptor is cleared from the descriptor
+    /// and may be assigned to new descriptors that are later added to the table.
+    /// Returns an owned pointer to the CompatDescriptor if the descriptor was found
+    /// in the table and removed, and NULL otherwise.
+    ///
+    /// NOTE: this will not unref the descriptor and you should do so manually.
+    #[no_mangle]
+    pub unsafe extern "C" fn descriptortable_remove(
```

Comment:
The returned type `CompatDescriptor` doesn't have a reference count. Maybe the comment should say it should be freed with `compatdescriptor_free()`?

Also this function doesn't ever return NULL, but probably should following the earlier comment about `process_deregisterCompatDescriptor()`.

Reply:
Done